gcore/src/error.rs

```diff
@@ -0,0 +1,56 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pub use gear_core_errors::{ExtError, MemoryError, MessageError, TerminationReason};
+
+#[cfg(feature = "codec")]
+mod sys {
+    extern "C" {
+        pub fn gr_error(data: *mut u8);
+    }
+}
+
+#[must_use]
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SyscallError {
+    len: u32,
+}
+
+impl SyscallError {
+    pub fn into_result(self) -> Result<(), ExtError> {
+        if self.len == 0 {
+            Ok(())
+        } else {
+            #[cfg(feature = "codec")]
+            unsafe {
+                use alloc::vec;
+                use codec::Decode;
+
+                let mut data = vec![0; self.len as usize];
+                sys::gr_error(data.as_mut_ptr());
+                Err(ExtError::decode(&mut data.as_slice()).expect("error decoded successfully"))
+            }
```

Comment:
maybe wrap this with some safe function?

Reply:
it would be really nice to have one with some doc-comments pointing to the fact, that we get last error and bla bla

Reply:
I don't think we need to introduce an another function for so small function. I'll just leave comment on `unsafe` block

Reply:
well it's commonly recommended pattern, better follow commonly accepted rules.

Reply:
What pattern?

Reply:
https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html

```md
To isolate unsafe code as much as possible, it’s best to enclose unsafe code within a safe abstraction and provide a safe API, which we’ll discuss later in the chapter when we examine unsafe functions and methods. Parts of the standard library are implemented as safe abstractions over unsafe code that has been audited. Wrapping unsafe code in a safe abstraction prevents uses of unsafe from leaking out into all the places that you or your users might want to use the functionality implemented with unsafe code, because using a safe abstraction is safe.
```